### PR TITLE
Chat gpt attempt questions

### PR DESCRIPTION
**Pull request by tmcdonnell2**
New and improved version of ChatGPT posting to Discord:
- ~~Longer wait time before timeout~~
- Break apart responses into max 2000 character limit
- Included test cases for testing breaking up messages

Also attempts questions posted in the question forum (attempts due to fear of timeouts and the classic "I'm sorry, as a LLM I can't...")

Resolves Together-Java/TJ-Bot/issues/814,
~~Resolves Together-Java/TJ-Bot/issues/839,~~
Resolves Together-Java/TJ-Bot/issues/840

### CODE REVIEW

**Comment by SquidXTV** [comment]
You request reviews but still have some checks that were not successful?
I would fix these first and then request reviews

**Comment by SquidXTV** [comment]
Also if I remember correctly your pull request is working on some issues.
You might want to link these issues in the pr so they get closed automatically 
if pushed.

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a class to break up long text blocks into smaller blocks which work with Discord's
+ * API. Initially constructed to break apart text from AI text generation APIs.
+ */
+public class AIResponseParser {
+
+    private AIResponseParser() {}
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:17]: in other utility classes we also throw an error, do that here as well

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a class to break up long text blocks into smaller blocks which work with Discord's
+ * API. Initially constructed to break apart text from AI text generation APIs.
+ */
+public class AIResponseParser {
+
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:16]: remove this line maybe?

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a class to break up long text blocks into smaller blocks which work with Discord's
+ * API. Initially constructed to break apart text from AI text generation APIs.
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:13]:
```suggestion
 * Represents a class to partition long text blocks into smaller blocks which work with Discord's
 * API. Initially constructed to partition long responses from AI text generation APIs.
```

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+     * @return An optional array potentially holding the original response split up into shorter
+     *         than {@value RESPONSE_LENGTH_LIMIT} length pieces.
+     */
+    public static Optional<String[]> parse(String response) {
+        String[] aiResponses;
+        if (response.length() > RESPONSE_LENGTH_LIMIT) {
+            logger.warn(
+                    "Response from AI was longer than allowed limit. "
+                            + "The answer was cut up to max {} characters length messages",
+                    RESPONSE_LENGTH_LIMIT);
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:37]: y warn?

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            aiResponses = breakupAiResponse(response);
+        } else {
+            aiResponses = new String[] {response};
+        }
+
+        return Optional.of(aiResponses);
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:43]: y return an optional? from my understanding it always returns an string array

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    }
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:44]:
```suggestion
     * @return An array potentially holding the original response split up into shorter
     *         than {@value RESPONSE_LENGTH_LIMIT} length pieces.
     */
    public static String[] parse(String response) {
        String[] aiResponses = new String[] {response};
        if (response.length() > RESPONSE_LENGTH_LIMIT) {
            aiResponses = breakupAiResponse(response);
        }

        return aiResponses;
    }
```

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java [hunk 7]
```diff
@@ -37,17 +37,34 @@ public ChatGptService(Config config) {
...
+        // Sending the system setup message to ChatGPT.
+        openAiService.createChatCompletion(systemSetupRequest);
     }
 
     /**
      * Prompt ChatGPT with a question and receive a response.
-     * 
+     *
      * @param question The question being asked of ChatGPT. Max is {@value MAX_TOKENS} tokens.
+     * @return response from ChatGPT as a String.
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java:63]:
```suggestion
     * @return partitioned response from ChatGPT as a String array.
```

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java [hunk 8]
```diff
@@ -59,15 +76,18 @@ public Optional<String> ask(String question) {
...
-            return Optional.ofNullable(openAiService.createChatCompletion(chatCompletionRequest)
+
+            String response = openAiService.createChatCompletion(chatCompletionRequest)
                 .getChoices()
                 .get(0)
                 .getMessage()
-                .getContent());
+                .getContent();
+
+            return AIResponseParser.parse(response);
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java:90]: no, u can't just remove the optional from there and return it in ur parse method

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java [hunk 9]
```diff
@@ -90,9 +91,17 @@ private void handleHelpThreadCreated(ThreadChannel threadChannel) {
         SERVICE.schedule(createMessages, 5, TimeUnit.SECONDS);
     }
 
+    private RestAction<Message> createAIResponse(ThreadChannel threadChannel) {
+        RestAction<Message> originalQuestion =
+                threadChannel.retrieveMessageById(threadChannel.getIdLong());
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java:96]: y u used the thread channel id to get the first message? is that valid?

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java [hunk 10]
```diff
@@ -90,9 +91,17 @@ private void handleHelpThreadCreated(ThreadChannel threadChannel) {
...
+        return originalQuestion.flatMap(
+                message -> helper.constructChatGptAttempt(threadChannel, message.getContentRaw()));
+    }
+
     private RestAction<Message> createMessages(ThreadChannel threadChannel) {
         return sendHelperHeadsUp(threadChannel).flatMap(Message::pin)
-            .flatMap(any -> helper.sendExplanationMessage(threadChannel));
+            .flatMap(any -> helper.sendExplanationMessage(threadChannel))
+            .flatMap(any -> threadChannel.retrieveMessageById(threadChannel.getIdLong()));
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java:104]: y retrieve the message here and not do anything with it?

**Review by Taz03 — CHANGES_REQUESTED** [review]
maybe make the test files `.txt`

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java (hunk 2 above)
Comment (tmcdonnell2) [reply application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:16]: Which line number are you talking about?

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java (hunk 5 above)
Comment (tmcdonnell2) [reply application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:43]: That it does. I think I am still a bit shaky on when to use an Optional over just the specific value. 

I think this also made working with the the calling function easier since it returns an optional.

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java (hunk 6 above)
Comment (tmcdonnell2) [reply application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:44]: Oh I see, remove the debug statement. Got it.

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java (hunk 8 above)
Comment (tmcdonnell2) [reply application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java:90]: Not sure what is going on here. Do you mean the function call? It originally returned an optional string array. I now have line 90 as:
```java
return Optional.of(AIResponseParser.parse(response));
```
Is this what you mean?

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java (hunk 10 above)
Comment (tmcdonnell2) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java:104]: Looks like old code. Thanks for pointing it out. Removed.

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java (hunk 9 above)
Comment (tmcdonnell2) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java:96]: It works if that's what you mean. Calling threadChannel.getParentMessage causes an error:
```Parent of this thread is not a MessageChannel. Parent: net.dv8tion.jda.internal.entities.ForumChannelImpl@57f68dd7```

I think I found some documentation somewhere that states that using the thread channels id will return the first message. Just tried to look for it and couldn't find it.

Is there another method I should be using?

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java (hunk 2 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:15]: 16 the empty line

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java (hunk 8 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java]: no, go by the logic, previously Optional.ofNullable() was wrapped on the return statement, which now u store in ur `response` variable. So that means it can be null. So u'll have to handle that case in either ur parse method but i'd recommend just return an empty optional if that's the case and not invoke the parse method at all

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+     * Parses the response generated by AI. If response is longer than
+     * {@value RESPONSE_LENGTH_LIMIT}, then breaks apart the response into suitable lengths for
+     * Discords API.
+     * 
+     * @param response The response from the AI which we want to send over Discord.
+     * @return An array potentially holding the original response split up into shorter than
+     *         {@value RESPONSE_LENGTH_LIMIT} length pieces.
+     */
+    public static String[] parse(String response) {
+        String[] aiResponses = new String[] {response};
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:28]: rename this to `partedResponse`

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        if (response.length() > RESPONSE_LENGTH_LIMIT) {
+            aiResponses = breakupAiResponse(response);
+        }
+
+        return aiResponses;
+    }
+
+    private static String[] breakupAiResponse(String response) {
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/AIResponseParser.java:36]: rename this to `partitionAiResponse`

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java (hunk 8 above)
Comment (tmcdonnell2) [reply application/src/main/java/org/togetherjava/tjbot/features/chaptgpt/ChatGptService.java]: I just pushed what I think is the change you are requesting. This is now outdated.

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpSystemHelper.java [hunk 13]
```diff
@@ -131,6 +151,90 @@ private RestAction<Message> sendExplanationMessage(GuildMessageChannel threadCha
...
+                && questionTitle.length() < MIN_QUESTION_LENGTH) {
+            return Optional.empty();
+        }
+
+        questionBuilder.append(questionTitle).append(" ");
+        if (originalQuestion.length() > MAX_QUESTION_LENGTH - questionBuilder.length()) {
+            originalQuestion =
+                    originalQuestion.substring(0, MAX_QUESTION_LENGTH - questionBuilder.length());
+        }
+
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpSystemHelper.java:213]:
```suggestion
        originalQuestion =
                    originalQuestion.substring(0, Math.min(MAX_QUESTION_LENGTH - questionBuilder.length(), originalQuestion.length()));
```

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpSystemHelper.java [hunk 14]
```diff
@@ -131,6 +151,90 @@ private RestAction<Message> sendExplanationMessage(GuildMessageChannel threadCha
...
+        StringBuilder tagBuilder = new StringBuilder();
+        int stringLength = questionBuilder.length();
+        for (ForumTag tag : threadChannel.getAppliedTags()) {
+            String tagName = tag.getName();
+            stringLength += tagName.length();
+            if (stringLength > MAX_QUESTION_LENGTH) {
+                break;
+            }
+            tagBuilder.append(String.format("%s ", tagName));
+        }
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpSystemHelper.java:225]: i dont think this'll help, have u tested it?

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpSystemHelper.java [hunk 15]
```diff
@@ -131,6 +151,90 @@ private RestAction<Message> sendExplanationMessage(GuildMessageChannel threadCha
...
+        Optional<String> questionOptional = prepareChatGptQuestion(threadChannel, originalQuestion);
+        Optional<String[]> chatGPTAnswer;
+
+        String question = questionOptional.orElseThrow();
+        logger.debug("The final question sent to chatGPT: {}", question);
+        chatGPTAnswer = chatGptService.ask(question);
+
+        if (chatGPTAnswer.isEmpty()) {
+            logger.warn("Something went wrong while trying to communicate with the ChatGpt API");
+            return useChatGptFallbackMessage(threadChannel);
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpSystemHelper.java:177]: maybe do something like this also in case of `questionQptional` being empty, instead of just throwing random errors in code

File: application/src/test/java/org/togetherjava/tjbot/features/chatgpt/AIResponseParserTest.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        for (String response : responses) {
+            Assertions.assertTrue(response.length() <= AI_RESPONSE_CHARACTER_LIMIT,
+                    "Response length is NOT within character limit: " + response.length());
+            logger.warn("Response length was: {}", response.length());
+        }
+    }
+
+    private void toLog(String[] responses) {
+        for (String response : responses) {
+            logger.warn(response);
```
Comment (Taz03) [inline application/src/test/java/org/togetherjava/tjbot/features/chatgpt/AIResponseParserTest.java:45]: use logger.info

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpSystemHelper.java (hunk 14 above)
Comment (tmcdonnell2) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpSystemHelper.java:225]: I have. All the tags of the question are appended to the beginning of the question. If you don't, there might not be enough context in the title or question to answer a question. Since we are a Java discord server, people usually don't refer directly to Java for simple questions. 

So if some asks:
```
Title: "Help" 
Question:"How do I print to the console/command line?"
```

WE all know that is meant to be Java but ChatGPT states "Sure! What language are you using?" or "Ok. Here is an example in python". But one thing people are relatively good at is clicking the right tags to append to their question (in fact isn't have at least one a requirement?).

**Review by Taz03 — DISMISSED** [review]

**Comment by tmcdonnell2** [comment]
When this is pulled, is it possible to have it running on the test server? I am sure there is some edge cases that will come up with more testing/questions. The only thing might be to add a way to save the response from the AI so we can test it again (given that AI usually don't give the same responses twice).

**Comment by Taz03** [comment]
@Together-Java/moderators

**Review by SquidXTV — CHANGES_REQUESTED** [review]

**Comment by SquidXTV** [comment]
can I ask why this literally deletes the website?

**Comment by tmcdonnell2** [comment]
> can I ask why this literally deletes the website?

I was trying to merge with develop to keep the branch up to date. I think they removed the website from this repository and gave the website its own repository. I am just trying to merge the changes. I ultimately regret trying haha.

**Comment by SquidXTV** [comment]
it also shows that your pr is adding the JShell stuff which is simply not true

**Comment by tmcdonnell2** [comment]
> > > can I ask why this literally deletes the website?
> 
> > 
> 
> > I was trying to merge with develop to keep the branch up to date. I think they removed the website from this repository and gave the website its own repository. I am just trying to merge the changes. I ultimately regret trying haha.
> 
> 
> 
> ah ok, but couldnt you not just fetch/pull from develop again, so yours is up-to-date?

I believe I have. A lot of merge conflicts emerge that despite fixing can't be merged into the branch? I am very confused. We are trying to reconstruct this branch in another branch. I am wondering if a roll back would be better?

**Comment by tmcdonnell2** [comment]
> it also shows that your pr is adding the JShell stuff which is simply not true

Again - because I am trying to merge in from the develop branch. I was just trying to keep it updated with the main branch. Not trying to get credit or anything like that. 

I thought keeping the branch as updated with the main branch was a good idea? Or is that not the way to go?

**Comment by SquidXTV** [comment]
> Again - because I am trying to merge in from the develop branch. I was just trying to keep it updated with the main branch. Not trying to get credit or anything like that.
yeah I wasnt saying that you want credits for that lol
I am just saying that you prob added those files yourself instead of doing fetch/pull

> I thought keeping the branch as updated with the main branch was a good idea? Or is that not the way to go?
I think you would even need to fetch/pull new updates before being able to merge

**Comment by surajkumar** [comment]
yeah the PR is fudged so my advice is to create a new branch and recreate the PR

**Comment by SquidXTV** [comment]
> yeah the PR is fudged so my advice is to create a new branch and recreate the PR

is it possible that you save your progress and completely roll back your branch + fetch/pull and then add your stuff to it
